### PR TITLE
Add support for RISC-V AIA IMSIC

### PR DESCRIPTION
**Pull request by D3boker1**
**RISC-V AIA APLIC support:**

- Change the Bao's interrupt reservation mechanism;
- Add MSI Mode support in (V)APLIC driver;
- IMSIC driver.
- VIMSIC driver

**Software Versions:**

QEMU - v7.1.0
Linux - https://github.com/avpatel/linux/tree/riscv_aia_v1
OpenSBI - 1.2

### CODE REVIEW

**Comment by josecm** [comment]
@D3boker1 before starting the review I'd ask you to: 

1. Resolve the conflicts with the main branch;
2. Make the history a little more granular, so we can better understand the structure of the introduced chages.

**Comment by josecm** [comment]
@ninolomata @DavidMCerdeira I believie this is ready to be merged. Please add your reviews.

@D3boker1 it would be great if you could also add your approval after my changes.

**Review by josecm — DISMISSED** [review]

**Comment by D3boker1** [comment]
Hi @josecm!

I have tested these changes on [AlSaqr SoC](https://github.com/AlSaqr-platform/he-soc/tree/zdl/exp-path) (dual-core version), and everything went smoothly.
It looks great to me!

File: src/arch/riscv/interrupts.c [hunk 1]
```diff
@@ -68,15 +78,34 @@ void interrupts_arch_enable(irqid_t int_id, bool en)
...
+            csrs_sip_clear(SIP_SSIP);
+            break;
+        case IRQ_S_TIMER:
+            interrupts_handle(irqc_timer_int_id);
+            break;
+        case IRQ_S_EXT:
+            irqc_handle();
+            break;
+        default:
+            break;
```
Comment (DavidMCerdeira) [inline src/arch/riscv/interrupts.c:97]: Shouldn't this print warning?

File: src/arch/riscv/irqc/aia/imsic.c [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * For now we only support 1 guest file per hart.
+ * Should I remove the guest_file from the API?
+ */
+void imsic_inject_pend(size_t guest_file, irqid_t intp_id)
+{
+    UNUSED_ARG(guest_file);
+
+    // vcpu->regs.hstatus = (guest_file << HSTATUS_VGEIN_OFF);
```
Comment (DavidMCerdeira) [inline src/arch/riscv/irqc/aia/imsic.c:79]: remove comment

**Review by DavidMCerdeira — DISMISSED** [review]

File: src/arch/riscv/irqc/aia/inc/imsic.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    uint32_t seteipnum_le;
+    uint32_t seteipnum_be;
+} __attribute__((__packed__, aligned(0x1000ULL)));
+
+struct imsic_global_hw {
+    struct imsic_intp_file_hw s_file;
+} __attribute__((__packed__, aligned(0x1000ULL)));
+
+/**
+ * @brief Initilaizes the IMSIC
```
Comment (ninolomata) [inline src/arch/riscv/irqc/aia/inc/imsic.h:34]: typo in Initializes

File: src/arch/riscv/irqc/aia/inc/imsic.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+ *        interrupt file. It sets every intp as triggerable, disables all interrupts, enables
+ *        interrupt delivery, and maps the S-lvl interrupt file in memory.
+ *
+ */
+void imsic_init(void);
+
+/**
+ * @brief Check if a given interrupt is pending for the cpu that calls the function
+ *
+ * @param intp_id the inetrrupt to check
```
Comment (ninolomata) [inline src/arch/riscv/irqc/aia/inc/imsic.h:46]: typo interrupt

File: src/arch/riscv/interrupts.c (hunk 1 above)
Comment (D3boker1) [reply src/arch/riscv/interrupts.c:98]: Agree!
Please take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/interrupts.c#L97)

File: src/arch/riscv/interrupts.c [hunk 5]
```diff
@@ -92,13 +121,18 @@ void interrupts_arch_handle(void)
             // WARNING("unkown interrupt");
```
[the remark this answers is not in the document]
Comment (D3boker1) [reply src/arch/riscv/interrupts.c]: Take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/interrupts.c#L122)

File: src/arch/riscv/irqc/aia/imsic.c (hunk 2 above)
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/imsic.c]: done

File: src/arch/riscv/irqc/aia/inc/aplic.h [hunk 6]
```diff
@@ -247,6 +263,14 @@ void aplic_set_target_hart(irqid_t intp_id, cpuid_t hart);
  */
 uint8_t aplic_get_target_prio(irqid_t intp_id);
 
+/**
+ * @brief Return the EIID of a given interrupt
```
[the remark this answers is not in the document]
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/inc/aplic.h]: done

File: src/arch/riscv/irqc/aia/imsic.c [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+void imsic_handle(void)
+{
+    uint32_t intp_identity = (uint32_t)(csrs_stopei_read() >> STOPEI_EEID);
+
+    if (intp_identity != 0) {
+        enum irq_res res = interrupts_handle(intp_identity);
+        if (res == HANDLED_BY_HYP) {
+            /** Write to STOPEI to clear the interrupt */
```
[the remark this answers is not in the document]
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/imsic.c]: Take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/irqc/aia/imsic.c#L95)

File: src/arch/riscv/irqc/aia/inc/aplic.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/inc/aplic.h]: Take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/irqc/aia/inc/aplic.h#L267)

File: src/arch/riscv/irqc/aia/inc/irqc.h [hunk 8]
```diff
@@ -10,14 +10,38 @@
...
+#if (IRQC == APLIC)
+#define IRQC_MAX_INTERRUPT_HANDLERS IRQC_MAX_INTERRUPT_LINES
+#elif (IRQC == AIA)
+/**
+ * Why IRQC_TIMR_INT_ID and not IRQC_SOFT_INT_ID? With the AIA specification, the software interrupt
+ * is now delivered through the IMSIC, which means that the target hart will see it as an external
+ * interrupt. Thus, the total number of interrupts is the maximum number of interrupts supported by
+ * aplic, the timer interrupt, the maximum number of interrupts supported by imsic, and one to
+ * support/keep "<" logic.
+ */
```
[the remark this answers is not in the document]
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/inc/irqc.h]: This text was useful in the first version of IMSIC support. With the current version, we can remove it. Take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/irqc/aia/inc/irqc.h#L26)

File: src/arch/riscv/irqc/aia/inc/irqc.h [hunk 9]
```diff
@@ -26,34 +50,109 @@ static inline void irqc_init()
...
+            return;
+        }
+#endif
+        aplic_set_sourcecfg(pintp_id, HYP_IRQ_SM_EDGE_RISE);
+        aplic_set_enbl(pintp_id);
+        aplic_set_target_hart(pintp_id, cpu()->id);
+#if (IRQC == AIA)
+        aplic_set_target_eiid(pintp_id, msi_id);
+        // 0 means is for hypervisor use
+        aplic_set_target_guest(pintp_id, 0);
```
[the remark this answers is not in the document]
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/inc/irqc.h]: done. Take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/irqc/aia/inc/irqc.h#L97)

File: src/arch/riscv/irqc/aia/vaplic.c [hunk 10]
```diff
@@ -44,9 +46,14 @@ static inline cpuid_t vaplic_vcpuid_to_pcpuid(struct vcpu* vcpu, vcpuid_t vhart)
 
 static uint32_t vaplic_get_domaincfg(struct vcpu* vcpu);
 static uint32_t vaplic_get_target(struct vcpu* vcpu, irqid_t intp_id);
+#if (IRQC == APLIC)
+
```
[the remark this answers is not in the document]
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/vaplic.c]: done. Take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/irqc/aia/vaplic.c#L49)

File: src/arch/riscv/irqc/aia/inc/imsic.h (hunk 3 above)
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/inc/imsic.h]: Done. Take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/irqc/aia/inc/imsic.h#L34)

File: src/arch/riscv/irqc/aia/inc/imsic.h (hunk 4 above)
Comment (D3boker1) [reply src/arch/riscv/irqc/aia/inc/imsic.h]: Done. Take a look [here](https://github.com/D3boker1/bao-hypervisor/blob/bd10f0347eefad0e3f21652d51d09c8eb40979be/src/arch/riscv/irqc/aia/inc/imsic.h#L46).

**Comment by DavidMCerdeira** [comment]
@D3boker1 Please rebase this with main

**Comment by josecm** [comment]
@DavidMCerdeira I've rebased this on main. I think the PR is ready to go.

@ninolomata could you add your final approval after @D3boker1 changes?

**Review by josecm — APPROVED** [review]

**Review by ninolomata — APPROVED** [review]

**Review by DavidMCerdeira — APPROVED** [review]

**Pull request merged**